### PR TITLE
Set zero as grad in initialize of using placeholder parameter

### LINKED ISSUE #1799: I got NaN when only using shape placeholder.
I found that a gradient in link is not updated when using shape placeholder.
An example is as follows. You will get NaN in the last print function.:
```python
import chainer
from chainer import links as L
from chainer import functions as F
from chainer import Variable
import numpy as np

l = L.Linear(None, 100)

x = Variable(np.zeros((10, 20)).astype("f"))

o = chainer.optimizers.SGD()

o.setup(l)
l.cleargrads()
y = l(x)
print(l.W.data.shape)

y.grad = np.zeros_like(y.data) + 1.0
y.backward(retain_grad=True)

print(l.W.data)
o.update()
print(l.W.data)
```
I tested on the recent chainer. I suspect that it may arise from the reference  of the optimizer but I don't have any detailed specifications. Do you have any ideas?

### PR DESCRIPTION
**Pull request by okuta**
Fix #1799
This bug is critical.
If we use shape placeholder, NN learning will fail. It is difficult for users to notice this bug.

But, this PC is like a work-around. Please tell me about good design.

### CODE REVIEW

File: tests/chainer_tests/optimizers_tests/test_optimizers_by_linear_model.py [hunk 1]
```diff
@@ -189,7 +203,8 @@ def create(self):
 
 
 @testing.parameterize(*testing.product({
-    'dtype': [numpy.float16, numpy.float32, numpy.float64]
+    'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    'use_placeholder': [True, False],
```
Comment (delta2323) [inline tests/chainer_tests/optimizers_tests/test_optimizers_by_linear_model.py]: Why we test placeholders for some specific optimizers only?

**Comment by delta2323** [comment]
Could you add the failed case in #1799 so that the issue is surely resolved.

**Comment by delta2323** [comment]
> Please tell me about good design.

One alternative is to call cleargrad (or diractly fill grad attribute with nan).

**Comment by delta2323** [comment]
@okuta Should we close this PR and work on #1811?

**Comment by okuta** [comment]
Yes, Please see #1811 .

**Pull request closed without merging**